Issue 12544  Sum(cos(1/n) - 1, (n, 1, oo)).is_convergent() issue

maurogaravello opened the pull request:
The series of `cos(1/n) - 1` is convergent, but the
`is_convergent()` method return false.
To fix this issue, a function for finding the asymptotic behavior
is implemented.

The changes are in `summation.py`.
A test is added.

"Fixes #12544 "

maurogaravello committed:
1

maurogaravello committed:
2

maurogaravello committed:
3

maurogaravello committed:
Sum(cos(1/n) - 1, (n, 1, oo)).is_convergent() issue 12544

The series of cos(1/n) - 1 is convergent, but sympy returns false to the
is_convergent() method
I think that the problem is in the line 104 of solveset.py.
Sympy can't invert the function log(x)/x. In this case the method returns as 's' the value '0',
i.e. the rhs of equation log(x)/x=0. So s.intersection(domain) has not sense in this situation.
I added a test to check whether the result is an inversion. Probably it is not the best way,
but it seems to work.

maurogaravello committed:
Sum(cos(1/n) - 1, (n, 1, oo)).is_convergent() issue 12544

The series of cos(1/n) - 1 is convergent, but the
is_convergent() method return false.
To fix this issue, a function for finding the asymptotic behavior
is implemented.

The changes are in summation.py.
A test is added.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -1098,3 +1099,35 @@ def eval_sum_hyper(f, i_a_b):
+    return either the asymptotic expression or the original one
+    '''
+    from sympy import simplify
+    if type(expr) is Mul:
```
Use `isinstance` to check types.

maurogaravello replied:
Done

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -1098,3 +1099,35 @@ def eval_sum_hyper(f, i_a_b):
+            return s.args[0]
+        else:
+            return expr
+    except:
```
Don't use bare excepts. Catch only the exception you are expecting.

maurogaravello replied:
I put `except Exception:`. I am not sure if it is ok. I tried with `except NotImplementedError:`, but some tests failed.

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -1098,3 +1099,35 @@ def eval_sum_hyper(f, i_a_b):
         return Piecewise(res, (old_sum, True))
+
+def asymptotic(expr, symbol):
+    '''Tries to perform the asymptotic of expr when
```
asymptotic expansion?

maurogaravello replied:
Not always is an expansion. Probably 'asymptotic bahavior' is better

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -6,7 +6,7 @@
 from sympy.core.singleton import S
 from sympy.core.symbol import Dummy, Wild, Symbol
-from sympy.core.add import Add
+from sympy.core.add import Add, Mul
```
Mul should come from sympy.core.mul.

maurogaravello replied:
Done

maurogaravello committed:
Done the requested changes

maurogaravello commented:
In the case log(n/(n+1)) the method asymptotic does not work correctly. I will try to fix.

maurogaravello committed:
Improved asymptotic method

maurogaravello committed:
Modified asymptotic() function. Now the doctests pass.

maurogaravello commented:
@asmeurer What do you think with this PR?

maurogaravello committed:
Asymptotic behavior only in order

Modified sympy/concrete/summations.py in such a way the
asymptotic behavior of 'sequence_term' applies only to the
'order' variable.
